### DISCUSSION TITLE
EMQX as main data storage?

### DISCUSSION THREAD

**Discussion by YehorFedorov** [emqx/emqx#6130] [Q&A] [answered]
Hello community)

I have a few questions about my idea how to use emqx in my future project and I want to hear your thoughts and answers)

My main question is: Is it a good idea to use EMQX (or mqtt generally) as a data storage?
I'm talking about retained messages. I'm planning to use only retained messages and don't use qos param at all.

When I tested my idea a few years ago, I used emqx 4.0.1 version and this worked well but in current versions I started to notice that emqx dropping connections, when client subscribe to wildcard (topic/#, for example) and there are 10000+ topics under that wildcard

And now I'm thinking, isn't it an anti-pattern to use only retained messages and emqx as a main data storage? Or mqtt should be used only as a transport layer?

And if not, is this solution scalable? If I would like to create cluster with 2 nodes would them sync retained storage with each other?

**Comment by zmstone** [comment]
Hi @YehorFedorov

I wouldn't argue it's anti-pattern, rather, we'd admit that EMQ X is not doing great in processing retained messages when:
1. The broker has limited RAM (even with the 'disc' option configured, all message have a copy in RAM)

~2. Subscribed topic can match large amount of messages~

2. When there are large number of messages to scan for wildcard subscriptions 

Our plan is to introduce off RAM on-disk storage,
and feed subscribers which match large amount of messages with throttled resource consumption.

Nonetheless, the connection drop is probably something worth investigating
if you can share more insights on the reason or maybe some logs (debug level would be great).

**Reply by YehorFedorov** [reply]
Hello @zmstone!
Thank you for a quick response!)

I'm happy to hear that this is already in your plans!
A few weeks ago I saw issue about connection drop and somebody from emqx team answered, that this bug will be fixed in v5.0+, if I'm not mistaking. But now I can't find this post :(

I'd be happy to help you investigate this problem and will provide logs and other info in few days (maybe logs would help me find resolved issue in github)

**Comment by coder4803** [comment]
@zmstone

> 2. Subscribed topic can match large amount of messages

It actually has a performance problem even if the wildcard subscription matches none of the retained messages (as described in #6014)

Additionally in our tests emqx performed very well when we sent one wildcard subscription that matched for 30k retained messages on the broker. The problem is that handling a wildcard subscription (probably finding those matches from the pool/cache) is quite slow.

So it is not a problem if you send a single wildcard subscription that matches a huge number of retained topics. But the problem is if you send huge number of wildcard subscriptions that do match none (or very few) retained topics (when you compare results to other brokers).

**Reply by zmstone** [reply]
Hi @coder4803 
Thank you for pointing that out.
I have updated the answer.
The current implementation is based on Erlang's builtin DB's pattern match, it may fallback to a full table scan for wildcards.
we'll have a look if there is a quick remedy for this.

**Comment by YehorFedorov** [comment, marked as answer]
Hi @zmstone !

I found related issue to my problem - https://giters.com/emqx/emqx/issues/5431

Here is my results:
```
OS: MacOS Big Sur v11.4
Environment: Docker version 20.10.5, docker-compose version 1.29.0

EMQX v4.0.1 and v4.3.10

100000 retained topics:
topic/(1-100000)/value => 'value'

Subscribe topic: 'topic/#'
```

Results:
EMQX v4.0.1 - everything works fine
EMQX v4.3.10 - emqx dropping connection on wildcard subscription

Attaching archive with docker-compose.yml, mnesia and logs dirs
[emqx-test.zip](https://github.com/emqx/emqx/files/7538233/emqx-test.zip)

**Reply by zmstone** [reply]
@YehorFedorov Thank you for the report.
We'll look into this ASAP.

**Reply by hjianbo** [reply]
First of all, the topic design may need to be optimized, as emqx-4.x and below implementations immediately deliver all matched messages to the client, which can cause high system stress peaks. And the client may not be able to handle it in time. We may have optimization plans for version 5.0 or 4.3.x.

**Reply by hjianbo** [reply]
Then, I checked the logs and found that EMQ X  broker had delivered over 90,000 messages, but received a disconnect request from the client, which is why the connection was closed.
<img width="1258" alt="image" src="https://user-images.githubusercontent.com/13825269/141782068-99f8ec2c-9388-4c6f-8b79-4eae0f180fcb.png">

**Reply by YehorFedorov** [reply]
Hi @HJianBo !

I'm gonna use homie convention for my project
Given topic design is just for testing purposes)

I apologise for incorrect logs. Didn't check emqx2 dir for logs 😢

log/emqx contains logs from emqx v4.0.1 which works fine

Attaching logs from emqx v4.3.10
[emqx2.zip](https://github.com/emqx/emqx/files/7538963/emqx2.zip)

If you need example with exact topic design which I'm planning to use, please notice me)

**Reply by zmstone** [reply]
Hi @YehorFedorov 
The subscriber is killed by an overload protection threshold.
The limit is by default 64MB,  you can try to increase the allowed heap size by changing `force_shutdown_policy` config.

```
2021-11-15T13:16:23.096895+00:00 [error] Process:          <0.2040.0> on node 'emqx_broker@127.0.0.1', Context:          maximum heap size reached, Max Heap Size:    10485760, Total Heap Size:  11567947, Kill:             true, Error Logger:     true, GC Info:          [{old_heap_block_size,1727361},{heap_block_size,8142745},{mbuf_size,1698058},{recent_size,806457},{stack_size,219},{old_heap_size,963531},{heap_size,2984661},{bin_vheap_size,4560},{bin_vheap_block_size,46422},{bin_old_vheap_size,0},{bin_old_vheap_block_size,46422}]
```

**Reply by YehorFedorov** [reply]
Thanks @zmstone @HJianBo !

**Reply by zmstone** [reply]
you are welcome @YehorFedorov 
did the fix work for you?

**Reply by YehorFedorov** [reply]
@zmstone yes!

**Comment by zmstone** [comment]
@coder4803 Could you check if it's the same situation for you?
i.e. the connection is killed by overload protection policy (which has a 64MB heap-size limit by default).

**Reply by coder4803** [reply]
@zmstone Heap-size limit is not the problem in my case (and no disconnect happens either). My problem is the slowness of handling/finding matches for a wildcard subscribe. I do thousands of wildcard subscriptions at once (each subscription typically matches 0-2 retained topics). In typical case there are 100k retained topics on the broker so it will take quite long time just to handle those subscriptions.

If I do only one subscription (like some/prefix/#) that matches 100k retained topics, it is actually handled (and messages are delivered) quite fast (no performance problems at all). But creating only one or few subscriptions is not an option in my case since it would subscribe huge number of non-interesting topics (which would cause another performance problem :) )

The pattern of typical subscription in my case is "some/prefix/+/other/postfix"

So in my case getting all subscribed (retained) messages takes quite a long time (when thousands of subscriptions are sent). Additionally it will consume all broker resources slowing down delivering other messages.

**Reply by zmstone** [reply]
Thank you for the detailed answer @coder4803.
Could you be more specific about the prefix part, is it the same for all (or most of) the subscribers?
or do you have a device ID kind of unique identifier before the `+` level.

**Reply by coder4803** [reply]
Yes, I am glad to share more details about the pattern.

Topic Pattern with problematic wildcard topic case is: "const1/const2/< var1 >/const3/+/< some/topic/structure >" where
- **const1**, **const2** and **const3** are always the same
- **< var1 >** has ~200 different values (this is like device id)
- **< some/topic/structure >** has ~30 different combinations (1-5 parts i.e. "test/event" or "my/very/first/test/value). This is like value id.
- There is a relation between **< var1 >** and **< some/topic/structure >**: each **< var1 >** value has same set of **< some/topic/structure >** postfixes (well there are few device types so the combination is not exactly the same for each but roughly).

In this case there are 200 * 30 = 6000 wildcard subscriptions. Do you think that ordering those topics in certain order would make handling wildcard subscriptions remarkable faster on broker side?

Edit: these subscriptions are made by monitoring applications and the subscription set is the same for all of them (at startup).

**Comment by joshemq** [comment]
Hi @YehorFedorov, would you be open to sharing more details about your project with me? I'm EMQ's marketing director in the US, and I'd appreciate the chance to interview you and maybe share your project with others if you'd be interested in some promotion. Would it be okay to contact you directly?

**Reply by YehorFedorov** [reply]
Greetings, @joshemq!
Sure, let's talk)
Contact me by email)

**Comment by ahapeter** [comment]
Hi @joshemq, any update on this feature? Is it **production ready** for now?!
My requirement are
- use retained feature as a kv store. Which support strong consistency?! (data always store in retained store before publish to client?!)
    - use [http API](https://docs.emqx.com/en/emqx/v5.8/admin/api-docs.html#tag/Retainer/paths/~1mqtt~1retainer~1message~1%7Btopic%7D/get) to get retained message (if needed)
- i will control at subscribe side, so, no too short wild-card subscribe (which cause much scan then match) allowed
